src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -82,153 +90,204 @@ public static bool IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLock
-                    return false;
+                    (var hasTransitiveDependencyChanged, var tmessage) = HasProjectTransitiveDependencyChanged(framework.CentralPackageVersions, transitiveDependenciesEnforcedByCentralVersions, transitiveDependencies);
+                    if (hasTransitiveDependencyChanged)
+                    {
+                        // lock file is out of sync
+                        invalidReasons.Add(tmessage);
+                    }
                 }
 
-                var queue = new Queue<Tuple<string, string>>();
-                var visitedP2PReference = new HashSet<string>();
-
-                foreach (var projectReference in framework.ProjectReferences)
+                // Validate all P2P references
+                foreach (var framework in project.RestoreMetadata.TargetFrameworks)
                 {
-                    if (visitedP2PReference.Add(projectReference.ProjectUniqueName))
+                    var projecttarget = project.TargetFrameworks.FirstOrDefault(
+                        t => EqualityUtility.EqualsWithNullCheck(t.FrameworkName, framework.FrameworkName));
+
+                    if (projecttarget == null)
                     {
-                        var spec = dgSpec.GetProjectSpec(projectReference.ProjectUniqueName);
-                        queue.Enqueue(new Tuple<string, string>(spec.Name, projectReference.ProjectUniqueName));
+                        // This should never be hit. A hit implies that project.RestoreMetadata.TargetsFrameworks and project.TargetsFrameworks are not the same.
+                        throw new Exception(string.Format(
+                                    CultureInfo.CurrentCulture,
+                                    Strings.PackagesLockFile_ProjectIsMissingRestoreMetadataTfms,
+                                    framework.FrameworkName.GetShortFolderName()
+                                    ));
+                    }
 
-                        while (queue.Count > 0)
-                        {
-                            var projectNames = queue.Dequeue();
-                            var p2pUniqueName = projectNames.Item2;
-                            var p2pProjectName = projectNames.Item1;
-                            var projectDependency = target.Dependencies.FirstOrDefault(
-                                dep => dep.Type == PackageDependencyType.Project &&
-                                StringComparer.OrdinalIgnoreCase.Equals(dep.Id, p2pProjectName));
-
-                            if (projectDependency == null)
-                            {
-                                // project dependency doesn't exist in lock file.
-                                return false;
-                            }
+                    var target = nuGetLockFile.Targets.FirstOrDefault(
+                        t => EqualityUtility.EqualsWithNullCheck(t.TargetFramework, framework.FrameworkName));
+
+                    if (target == null)
```

Comment:
Yes, because addition of new tfm to the project invalidates lock file. this could be one of the reasons. I modified logic in this PR to capture all the possible reasons why the lock file is invalid.